src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/ExternalProcessRunner.java

```diff
@@ -0,0 +1,103 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+@Slf4j
+@Component
+public class ExternalProcessRunner {
+    /**
+     * Executes the given command with given arguments.
+     * @param command the command to execute
+     * @param args arguments to be appended to the command. Make sure to pass your arguments in the correct order
+     * (e.g. if your options have values enter them as separate consecutive args).
+     * @return the output of the executed process as a List of Strings
+     * @throws ProcessNotExecutableException in the case of IOException or if the process is interrupted
+     * @throws ProcessFailedException if the process' exit code is not 0
+     */
+    public List<String> execute(String command, String... args) throws ProcessNotExecutableException,
+            ProcessFailedException {
+        if (StringUtils.isEmpty(command)) {
+            throw new IllegalArgumentException("command cannot be null");
+        }
+
+        List<String> commandWithArgs = new ArrayList<>();
+        commandWithArgs.add(command);
+        if (args != null && args.length > 0) {
+            commandWithArgs.addAll(Arrays.asList(args));
+        }
+        Process process;
+        ProcessBuilder pb = new ProcessBuilder(commandWithArgs);
+        // redirect process errors into process's input stream
+        pb.redirectErrorStream(true);
+        try {
+            process = pb.start();
+        } catch (IOException e) {
+            throw new ProcessNotExecutableException(e);
+        }
+
+        // gather output into a list of string for returning
+        List<String> processOutput = new ArrayList<>();
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
+            br.lines().forEach(processOutput::add);
+        } catch (IOException e) {
+            process.destroy();
+            throw new ProcessNotExecutableException(e);
+        }
+
+        int exitCode;
+
+        try {
+            exitCode = process.waitFor();
+        } catch (InterruptedException e) {
+            // we don't want to throw the InterruptedException from here but we want to retain the interrupted status
+            Thread.currentThread().interrupt();
+            throw new ProcessNotExecutableException(e);
+        } finally {
+            // always destroy the process
+            process.destroy();
+        }
+
+        // if the process fails we attach the output into the exception
+        if (exitCode != 0) {
+            String fullCommandString = String.join(" ", commandWithArgs);
+            String processOutputString = String.join("\n", processOutput);
+            String errorMsg = String.format("Failed to run command '%s' with output: \n %s", fullCommandString,
+                    processOutputString);
+            throw new ProcessFailedException(errorMsg, processOutput);
```

Comment:
This stores process output in the ErrorDeviation metadata. I would like to remove that as well - it is in a sense misuse of the metadata. Maybe just add some sort of separate container property in ProcessFailedException (which may be more clear option anyways)?

We have DeviationAwareException constructors that take a DeviationAware object as a parameter and copy ErrorDeviation and metadata from there - it feels likely that the process output could start to leak to the API response by mistake. I may have written a better and longer explanation here https://github.com/nordic-institute/X-Road/pull/344#discussion_r372260209

Reply:
Yes it was not supposed to even be there because the process output is already stringified to the error message. Will be removed.